Helper methods

brendanmckeown opened the pull request:
Let me know if you want to change anything with these methods.

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+		$value = null;
+		if (self::tableColumnExists($table, $field)) {
+			$query = new SQLQuery();
+			$query->setFrom($table)->setSelect([$field])->setWhere("ID = $id");
```
Would you mind casting this to an integer first? Unfortunately SS doesn't have parametrization yet. 2 other spots below too.

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+		$values = [];
+		if (self::tableColumnsExist($table, $fields)) {
+			$query = new SQLQuery();
+			$query->setFrom($table)->setSelect($fields)->setWhere("ID = $id");
```
Here.

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+				if ($i < $valuesCount - 1) $query .= ",";
+				$i++;
+			}
+			$query .= " WHERE ID = $id;";
```
And here.

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+	protected static function setRowValuesOnTable($table, array $values, $id) {
+		$queried = false;
+		if (self::tableColumnsExist($table, array_keys($values))) {
+			$query = "UPDATE $table SET";
```
SS might have something to quote the table name here. Not at PC to look.

patricknelson commented:
Nice work! Tons of great changes here, some stuff I've yet to to encounter too, so it should be useful later when the time comes. Just a few SQL things. That and I need to test it out. Also I should probably get in there and drop in some unit tests and setup Travis CI.

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+	 * @return boolean
+	 */
+	protected static function dropColumnsFromTable($table, array $columns) {
+		$queried = false;
```
Just a muse, but this is where array_intersect and array_keys is handy too. Start out getting the intersection, if it's empty, return false. It will be empty if there is no table or if the specified array of columns aren't found. Then, just iterate through that array and issue the  `ALTER` query on the (escaped) table and return true at the end (since we already know the array wasn't empty).

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+			if ($results) {
+				foreach ($results as $result) {
+					foreach ($fields as $field) {
+						$values[$field] = $result[$field];
```
I really wish SS supported (and we ran) PHP 5.5 so we could just use array_column() because this is tedious, being stuck in the past (supporting 5.3).

patricknelson replied:
Maybe http://php.net/manual/en/function.array-intersect-key.php would be useful here too :D

patricknelson commented on code/Migration.php:
```diff
@@ -14,4 +14,189 @@
+			$i = 0;
+			foreach ($values as $field => $value) {
+				if (is_string($value)) $value = "'" . Convert::raw2sql($value) . "'";
+				$query .= " $field = " . $value;
```
`$table` and `$field` should be escaped (too easy to have MySQL keywords as column names). Tip: Also you can just build an array and join it to generate the comma separated values, if you wanted (end result is the same so don't change it if you don't want to).

patricknelson commented:
One final thing: In our project I use the newer `[]` array syntax, but since this needs to support back to PHP 5.3 (due to the silverstripe-framework 3.x requirement) it will need to be updated to use the old `array()` syntax. Sorry for all the nagging :-/

That reminds me, so I just updated [my code here](https://github.com/patricknelson/silverstripe-migrations/commit/a8e125fbb245249a63bcc8e263132f694eb295fb) as well.

brendanmckeown commented:
Okay...I made some updates to these methods per your comments.  Casting all $id's to integers before use, escaping all values being used in a query, refactored some methods, and changed array syntax.

brendanmckeown commented:
Also, let me know once you add unit testing to this project and I will write some tests for these methods.

patricknelson commented:
Good idea -- I was already manually checking out each of the methods on my own. I think I might rip off some of the travis settings from the SS framework considering this needs to be compatible with that.

patricknelson commented:
For now, though -- go ahead and write your unit tests and place them under a new root `tests` folder. You'll need a dev dependency setup in `composer.json` as well, e.g.:

``` json
    "require-dev": {
        "phpunit/PHPUnit": "~3.7@stable"
    },
```

That should get you started at least (FYI though, you can run this in your existing SS installation). The dependency is there to be thorough.

patricknelson commented:
Ok ready, go ahead and pull/rebase whenever you'd like.

patricknelson commented:
Made more updates, so make sure you pull the latest from `master` and rebase your changes on that. Use the `->task` property on `MigrateTaskTest` class if you want to access the `MigrateTask` instance from within your test methods.
